File: javascript/word-count/word-count.js

export class Words {
  count(str) {
    return str
      .trim()
      .split(/[ \n\t]+/g)
      .map(word => word.toLowerCase())
      .reduce((words, current) => ({
        ...words,
        [current]: words.hasOwnProperty(current)
          ? words[current] + 1
          : 1
      }), {});
  }
}

File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> String.downcase()
    |> String.split(~r/[\s_,.:!?&@$%^]+/, trim: true)
    |> Enum.reduce(
      %{},
      fn word, acc ->
        case acc do
          %{^word => c} -> %{acc | word => c + 1}
          _ -> Map.put(acc, word, 1)
        end
      end
    )
  end
end
